tools/pal_inference.py

```diff
@@ -0,0 +1,471 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import copy
+import json
+import os
+import signal
+import warnings
+from dataclasses import asdict, dataclass
+from typing import Any, Callable, Dict, List, Optional
+
+import torch
+import torch.nn as nn
+import tqdm
+from datasets import load_dataset
+from transformers import AutoModel, AutoModelForCausalLM, AutoTokenizer
+from transformers.generation.utils import (LogitsProcessorList,
+                                           StoppingCriteriaList)
+from transformers.utils import logging
+
+logger = logging.get_logger(__name__)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='PAL Inference')
+    parser.add_argument('model',
+                        type=str,
+                        help='Path to the pre-trained LLM used for inference.')
+    parser.add_argument(
+        'out_dir',
+        type=str,
+        help='Name of the output folder where generated code snippets will be saved.'
+    )
+    parser.add_argument(
+        '--dataset',
+        default='gsm8k',
+        type=str,
+        help='Name of the dataset used for code generation (default: gsm8k).')
+    parser.add_argument(
+        '--max_length',
+        default=2048,
+        type=int,
+        help='Maximum input token length for the natural language description (default: 2048).'
+    )
+    parser.add_argument(
+        '--top_p',
+        default=0.8,
+        type=float,
+        help='Probability threshold to choose sample tokens during generation (default: 0.8).'
+    )
+    parser.add_argument(
+        '--temperature',
+        default=1,
+        type=float,
+        help='Temperature of token sampling during generation (default: 1).')
+    parser.add_argument(
+        '--time_out',
+        default=100,
+        type=float,
+        help='Maximum time allowed for executing generated code (default: 100).'
+    )
+    parser.add_argument(
+        '--verbose',
+        '-v',
+        action='store_true',
+        help='Print code error information when executing generated code (optional).'
+    )
+    parser.add_argument(
+        '--append',
+        '-a',
+        action='store_true',
+        help='Append output to the history results (optional).')
+    args = parser.parse_args()
+    return args
+
+
+@dataclass
+class GenerationConfig:
+    max_length: Optional[int] = None
+    top_p: Optional[float] = None
+    temperature: Optional[float] = None
+    do_sample: Optional[bool] = True
+    repetition_penalty: Optional[float] = 1.0
+
+
+@torch.inference_mode()
+def generate_interactive(
```

Comment:
if copied from web_demo.py and did not change anything, we can move the same code to `internlm/utils/inference.py`